refactor(compiler): implement block syntax in html ast

crisbeto opened the pull request:
⚠️Disclaimer⚠️ this PR implements syntax that is still in an open RFC. It will be adjusted once the RFC is closed.

Built on top of https://github.com/angular/angular/pull/50895. All the new code is in the second commit.

These changes implement the `BlockGroup` and `Block` AST nodes that will then be used to generate instructions based on the new syntax. A `BlockGroup` is a container for `Block` instances. The first block of a block is always implicit and required while any subsequent blocks are optional.

AndrewKushnir reviewed (approved):
@crisbeto looks great 👍

crisbeto committed:
refactor(compiler): implement block syntax in html ast

⚠️Disclaimer⚠️ this PR implements syntax that is still in an open RFC. It will be adjusted once the RFC is closed.

These changes implement the `BlockGroup` and `Block` AST nodes that will then be used to generate instructions based on the new syntax. A `BlockGroup` is a container for `Block` instances. The first block of a block is always implicit and required while any subsequent blocks are optional.

AndrewKushnir commented:
@crisbeto PR #50895 is now merged, so we can rebased and land this one :)

crisbeto commented:
Rebased to drop the previous commit. Re-running the presubmit.

crisbeto commented:
Caretaker note: passes presubmit, except for one unrelated failure.

dylhunn commented:
This PR was merged into the repository by commit b14a78ebb03c41532a34498dbe19924b60a1e82e.
